### PR TITLE
[SYCL] Unique stable name rebase

### PR DESCRIPTION
**Pull request by erichkeane**
This reverts our existing implementation, updates the library uses of this, and adds the community patch back:

8347ee428cf16512d5722d99111bd8d051668d28

### CODE REVIEW

**Comment by erichkeane** [comment]
Note that I noticed a bunch of check-all failures locally that i'm going to work through as well.  There shouldn't be much to review here, as the community reviewed the big patch, and this mostly just removes the old one and adds that.  There are a few library changes that are just changing the name.

**Review by s-kanaev — DISMISSED** [review]
RT changes LGTM

**Comment by erichkeane** [comment]
@AlexeySachkov : Can you look into the level-zero issue that showed up on the build-bots?  I look at it when it came up on the linux bot, but don't have a good idea on how to start debugging it.

**Comment by AlexeySachkov** [comment]
> @AlexeySachkov : Can you look into the level-zero issue that showed up on the build-bots? I look at it when it came up on the linux bot, but don't have a good idea on how to start debugging it.

@erichkeane, the failure is not directly related to level-zero, but it is related to "unnamed lambda" functionality. In Jenkins/Precommit there are ~18 failures and all of them because kernel wasn't found in device images.

I've used minimalistic (could be reduced more, of course) example to reproduce:
```
#include <CL/sycl.hpp>

constexpr sycl::specialization_id<int> test_id_1{42};

int main() {

  sycl::queue q;
  {
    sycl::buffer<double, 1> Buf{sycl::range{1}};
    q.submit([&](sycl::handler &cgh) {
      auto Acc = Buf.get_access<sycl::access::mode::read_write>(cgh);
      cgh.set_specialization_constant<test_id_1>(1);
      cgh.single_task<class Kernel1>([=](sycl::kernel_handler kh) {
        Acc[0] = kh.get_specialization_constant<test_id_1>();
      });
    });
    auto Acc = Buf.get_access<sycl::access::mode::read>();
    assert(Acc[0] == 1);
  }
}
```

```
$ clang++ -fsycl -fsycl-unnamed-lambda t.cpp
$ ./a.out
terminate called after throwing an instance of 'cl::sycl::exception'
  what():  The kernel bundle does not contain the kernel identified by kernelId. -30 (CL_INVALID_VALUE)
Aborted (core dumped)
```

`KernelInfo::getName()` returns an empty string. From what I see it seems like the template arguments of `KernelInfoData` structure (defined in `kernel_desc.hpp`) are different between the integration header and instantiation of that structure.

From integration header:
```
 // names of all kernels defined in the corresponding source
 static constexpr
 const char* const kernel_names[] = {
   "_ZTSZZ4mainENKUlRN2cl4sycl7handlerEE10000_clES2_EUlNS0_14kernel_handlerEE10000_
 template <> struct KernelInfoData<'_', 'Z', 'T', 'S', 'Z', 'Z', '4', 'm', 'a', 'i', 'n', 'E', 'N', 'K', 'U', 'l', 'R', 'N', '2', 'c'
 , 'l', '4', 's', 'y', 'c', 'l', '7', 'h', 'a', 'n', 'd', 'l', 'e', 'r', 'E', 'E', '1', '0', '0', '0', '0', '_', 'c', 'l', 'E', 'S',
 '2', '_', 'E', 'U', 'l', 'N', 'S', '0', '_', '1', '4', 'k', 'e', 'r', 'n', 'e', 'l', '_', 'h', 'a', 'n', 'd', 'l', 'e', 'r', 'E', 'E
 ', '1', '0', '0', '0', '0', '_'> {
```

During instantiation (I've commented `getName` method default implementation to trigger the error)
```
include/sycl/CL/sycl/handler.hpp:540:13: error: no member named
      'getName' in 'sycl::detail::KernelInfoData<'_', 'Z', 'T', 'S', 'Z', 'Z', '4', 'm', 'a', 'i', 'n', 'E', 'N', 'K', 'U', 'l', 'R',
      'N', '2', 'c', 'l', '4', 's', 'y', 'c', 'l', '7', 'h', 'a', 'n', 'd', 'l', 'e', 'r', 'E', 'E', '_', 'c', 'l', 'E', 'S', '2', '_',
      'E', 'U', 'l', 'N', 'S', '0', '_', '1', '4', 'k', 'e', 'r', 'n', 'e', 'l', '_', 'h', 'a', 'n', 'd', 'l', 'e', 'r', 'E', 'E', '_'>'
// Or: _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE_clES2_EUlNS0_14kernel_handlerEE_
```

**Comment by erichkeane** [comment]
Interesting.... The 'host' side is obviously wrong there, since it is a lambda without the unnamed-lambda spelling.  Could this be a missed "sycl_kernel" attribute?  One of the gotchas with the new mechanism is that the sycl_kernel attribute needs to be present on both sides.

**Review by AaronBallman — DISMISSED** [review]
LGTM

**Comment by erichkeane** [comment]
1 more patch :) @romanovvlad  please take a look at the warnings.cpp test from here: https://github.com/intel/llvm/pull/3835/commits/45d489eba0f971ea2cacc315eaa4faadbf92b7b2 to figure out if you want to make other different changes to your RT commits here.

Thanks!

**Comment by romanovvlad** [comment]
> 
> 
> 1 more patch :) @romanovvlad please take a look at the warnings.cpp test from here: [45d489e](https://github.com/intel/llvm/commit/45d489eba0f971ea2cacc315eaa4faadbf92b7b2) to figure out if you want to make other different changes to your RT commits here.
> 
> Thanks!

Will try to do tomorrow.

**Review by AaronBallman — COMMENTED** [review]
LGTM!

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -825,9 +825,8 @@ class __SYCL_EXPORT handler {
 
       range<Dims> AdjustedRange = NumWorkItems;
       AdjustedRange.set_range_dim0(NewValX);
-#ifdef __SYCL_DEVICE_ONLY__
       kernel_parallel_for_wrapper<NameWT, TransformedArgType>(Wrapper);
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/handler.hpp:828]: This is performed in both host and device, isn't it?
Why is that needed?

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -839,10 +838,9 @@ class __SYCL_EXPORT handler {
        // !DPCPP_HOST_DEVICE_OPENMP && !DPCPP_HOST_DEVICE_PERF_NATIVE &&       \
        // SYCL_LANGUAGE_VERSION >= 202001
     {
-#ifdef __SYCL_DEVICE_ONLY__
       (void)NumWorkItems;
       kernel_parallel_for_wrapper<NameT, TransformedArgType>(KernelFunc);
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/handler.hpp:842]: Same question here.

**Review by s-kanaev — APPROVED** [review]
Other than the questions, RT changes LGTM

File: sycl/include/CL/sycl/handler.hpp (hunk 1 above)
Comment (erichkeane) [reply sycl/include/CL/sycl/handler.hpp:828]: @romanovvlad  made these changes, but the root of it is that we're required to have the kernel instantiated on both sides now, since otherwise we cant tell whether a lambda is used to name a kernel on the host-side.

**Review by premanandrao — APPROVED** [review]
LGTM

**Comment by bader** [comment]
@erichkeane, this patch breaks "clang-12 + shared libraries" post-commit configuration - https://github.com/intel/llvm/runs/2738419633?check_suite_focus=true.
Could you take a look, please?

**Comment by erichkeane** [comment]
> 
> 
> @erichkeane, this patch breaks "clang-12 + shared libraries" post-commit configuration - https://github.com/intel/llvm/runs/2738419633?check_suite_focus=true.
> Could you take a look, please?

It appears the problem is in the library changes that @romanovvlad committed.  The 'simple' fix is to just push a commit that casts the parameters in each of these functions to 'void'.  I can prepare a patch to do that @bader, though I'm not sure how to submit it?

**Comment by bader** [comment]
Could you open a pull request with the patch you proposed, please? If possible, please, check locally that "self-build + shared lib" configuration passes compilation + lit tests. Unfortunately, this configuration is covered by post-commit only, so pre-commit doesn't validate that it's fixed.

**Comment by erichkeane** [comment]
Submitted a PR here: #3874

I don't have a self-build + shared-lib config available on my test environment, but the errors/changes seem quite simple.

**Comment by bader** [comment]
Interesting... I also see that `Clang :: Driver/sycl-device-lib.cpp` test from `check-clang` fails. Do you think it's related to the same problem and will be fixed by #3874?
+@mdtoguchi

**Comment by erichkeane** [comment]
I have seen that failure in my local repo for a few months now, I don't think it is related to this at all.

**Comment by bader** [comment]
I think it might be related to https://github.com/intel/llvm/issues/3630, but I thought that it's fixed by #3639. @jinge90, could you check, please?

**Comment by erichkeane** [comment]
So my mistake, that one no longer fails for me (but did for quite a while).  Even so, I don't see anything in this patch or the #3874 having the ability to change the behavior of that test at all.

FWIW, my environment's failures are now:  
Clang :: CodeGenCUDA/host-used-device-var.cu
  Clang :: CodeGenCUDA/unused-global-var.cu
  Clang :: Driver/amdgpu-openmp-toolchain.c
  Clang :: Driver/hip-include-path.hip

**Comment by bader** [comment]
I suspect that it might be that issue #3630 is not fixed, but your fix #3874 will hide it for CI. It's reproducible if you run clang tests w/o building sycl project. CI always builds sycl project before running clang lit tests. In this case it fails to build sycl project and run `check-clang`.
